kroxylicious-filters/kroxylicious-record-validation/src/main/java/io/kroxylicious/proxy/filter/schema/validation/topic/PerRecordTopicValidator.java

```diff
@@ -31,30 +33,34 @@ class PerRecordTopicValidator implements TopicValidator {
     }
 
     @Override
-    public TopicValidationResult validateTopicData(ProduceRequestData.TopicProduceData topicProduceData) {
-        return new PerPartitionTopicValidationResult(topicProduceData.name(), topicProduceData.partitionData().stream().collect(Collectors.toMap(
-                ProduceRequestData.PartitionProduceData::index, this::validateTopicPartition)));
+    public CompletionStage<TopicValidationResult> validateTopicData(ProduceRequestData.TopicProduceData topicProduceData) {
+        CompletableFuture<PartitionValidationResult>[] result = topicProduceData.partitionData().stream().map(this::validateTopicPartition)
+                .map(CompletionStage::toCompletableFuture)
+                .toArray(CompletableFuture[]::new);
+        return CompletableFuture.allOf(result).thenApply(unused -> {
+            Map<Integer, PartitionValidationResult> collect = Arrays.stream(result).map(CompletableFuture::join)
+                    .collect(Collectors.toMap(PartitionValidationResult::index, x -> x));
+            return new PerPartitionTopicValidationResult(topicProduceData.name(), collect);
+        });
     }
 
-    private PartitionValidationResult validateTopicPartition(ProduceRequestData.PartitionProduceData partitionProduceData) {
-        return new PartitionValidationResult(partitionProduceData.index(), validateRecords(partitionProduceData.records()));
-    }
-
-    private List<RecordValidationFailure> validateRecords(BaseRecords records) {
+    private CompletionStage<PartitionValidationResult> validateTopicPartition(ProduceRequestData.PartitionProduceData partitionProduceData) {
+        BaseRecords records = partitionProduceData.records();
         if (!(records instanceof MemoryRecords)) {
-            return List.of();
+            return CompletableFuture.completedFuture(new PartitionValidationResult(partitionProduceData.index(), List.of()));
         }
         int recordIndex = 0;
-        List<RecordValidationFailure> failures = new ArrayList<>();
-        for (MutableRecordBatch batch : ((MemoryRecords) records).batches()) {
-            for (Record record : batch) {
-                Result result = validator.validate(record);
-                if (!result.valid()) {
-                    failures.add(new RecordValidationFailure(recordIndex, result.errorMessage()));
+        CompletableFuture<List<RecordValidationFailure>> result = CompletableFuture.completedFuture(new ArrayList<>());
+        for (Record record : ((MemoryRecords) records).records()) {
+            int finalRecordIndex = recordIndex;
+            result = result.thenCompose(recordValidationFailures -> validator.validate(record).thenApply(result1 -> {
```

Comment:
is it the intent that validations are carried out sequentially? if so, why?  I get that we want results to be ordered.

Reply:
It is the intent, but just following the approach already present. Do you prefer to do it in batches?

Reply:
Ok.